developer_docs/syntactic_stage.md

```diff
@@ -0,0 +1,82 @@
+
+## Navigation and AST Creation
+
+The entry point for the class `SyntaxParser` is the function `parse`.
+This function is called from the constructor to examine a FST (Full Syntax Tree) object.
+The FST is the output of Python's `ast.parse` function with comments inserted using the [`pyccel.parser.extend_tree.extend_tree`](../pyccel/parser/extend_tree.py).
+Python's ast module discards all comments, so the function [`pyccel.parser.extend_tree.extend_tree`](../pyccel/parser/extend_tree.py) is needed to read the file and reinsert these comments at the correct location.
+
+The key line of the function `parse` is the call to `self._visit(self.fst)`.
+All elements of the tree must be visited.
+The `_visit` function internally calls a function named `_visit_X`, where `X` is the type of the object.
+This function must have the form:
+```python
+def _visit_ClassName(self, stmt):
+    ...
+```
+Each of these functions should internally call the `_visit` function on each of the elements of the object to obtain Pyccel AST nodes which can be combined to create a Pyccel AST node representing the current object.
+
+## Errors
+
+Error handling uses the classes in the file [pyccel.errors.errors](../pyccel/errors/errors.py).
+Errors in the syntactic stage should raise a `PyccelSyntaxError`.
+Where possible this should be done by accessing the `Errors()` singleton and calling the `report` function.
+This function takes several arguments (see docstring for more details).
+The most important arguments are:
+-   _message_ : Describe the issue that lead to the error
+
+-   _symbol_ : The Python ast object should be passed here. This object contains information about its position in the file (line number, column) which ensures the user can more easily locate their error
+
+-   _severity_ : The severity level must be one of the following:
+    -   _warning_ : An error will be printed but Pyccel will continue executing
+    -   _error_ : An error will be printed but Pyccel will continue executing the syntactic stage
+    -   _fatal_ : An error will be printed and Pyccel will stop executing. This level should rarely be needed in the syntactic stage as a failure in one function should not affect the execution of another. It is preferable to show the users all errors at once
+
+## Headers
+
+The headers (type declarations/openmp pragmas/etc) also have their own syntax which cannot be parsed by Python's ast module.
+The module [textx](http://textx.github.io/textX/stable/) is used to parse these statements.
+The files describing the _textx_ grammar are found in the folder [pyccel.parser.grammar](../pyccel/parser/grammar).
+From these files _textx_ generates instances of the classes found in the folder [pyccel.parser.syntax](../pyccel/parser/syntax).
+These instances can then be inserted into the annotated syntax tree.
```

Comment:
This is the first mention of _annotated syntax tree_. Do you think it should have been passively introduced above?

Reply:
Oops. It should be _abstract syntax tree_
Fixed 046b880d